cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -233,6 +236,104 @@ struct DivideChecked {
   }
 };
 
+template <typename T>
+inline T integer_power(KernelContext* ctx, T left, T right) {
+  if (right < 0) {
+    ctx->SetStatus(
+        Status::Invalid("integers to negative integer powers are not allowed"));
+  }
+  T result = 1;
+  if (left == 0 && right != 0) {
+    return 0;
+  }
+  while (true) {
+    if (right % 2) {
+      result *= left;
```

Comment:
Please do check the UBSAN CI failure. I think it's caused by signed integer overflow in this function. You can enable UBSAN check on your local host with `cmake -DARROW_USE_UBSAN=ON ....`.

Reply:
Weird - I've set the flag but don't get the overflow error locally, while [CI reports](https://github.com/apache/arrow/pull/9841/checks?check_run_id=2292038374#step:8:2054):
```
[ RUN      ] TestBinaryArithmeticIntegral/2.Power
/arrow/cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:256:10: runtime error: signed integer overflow: 2147483647 * 2147483647 cannot be represented in type 'int'
```

Reply:
It should work only by enabling this flag. Maybe you need to delete all files before running cmake again. I don't know.
To verify, you can build with `ninja -v` and check the command line if `-fsanitize=undefined` is there.
You can also check the CI job logs to see what flags it enabled in building.

Reply:
I got it working finally, thanks!
I think the issue is related to 16 bit multiplication as per comments in the Multiply kernel. I took the same approach and it seems to have help.